### PR TITLE
[Save Post Task Force] Prevent auto-saves after a user-initiated save and remove `SaveOnEditorExit` exception logging

### PR DESCRIPTION
**Pull request by ashiagr**
Internal Reference: pb3aDo-18j-p2

## Description

This PR adds a flag to prevent triggering “**auto-save**” events after a “**user-initiated**” save event.

This is because the “**user-initiated**”/ “**save on exit**” event captures the latest state of the UI (the source of truth) and any further updates received from `Gutenberg` after this point (and after [the title and content serialization fix](https://github.com/WordPress/gutenberg/pull/36072)) will represent the same state.

We can now remove the [SaveOnExitException](https://github.com/wordpress-mobile/WordPress-Android/blob/71240284d0f0d3aeb16a79416f45a394f7585e2b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L1763-L1775) logging in the `DEBUG` build because:

1. The ”**auto-save**”  events that get the content from the UI/ Gutenberg editor (see T1-T4) are queued as they're [separated by delays](https://github.com/wordpress-mobile/WordPress-Android/blob/71240284d0f0d3aeb16a79416f45a394f7585e2b/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt#L100); a “**user-initiated**” save (S) event may complete before the queue is empty. All pending “**auto-saves**” (see T3-T4) after  “**user-initiated**” save are prevented from starting by adding the flag above. This makes the [AutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L13-L17) condition redundant as we’re preventing any further “auto-saves”.

2. The “user-initiated” save (S) event can update the `PostModel` if a queued “auto-save” (T2) for a previous UI state completes before it. In this case, [NoAutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L23-L26) becomes redundant as the “user-initiated” save (S) had changes to be saved after “auto-save” (T2) while there was no auto-save pending due to 1 above.

   thread-state

~~Note: This PR is still in draft as it needs to target https://github.com/wordpress-mobile/WordPress-Android/pull/15508 which is currently failing + I had local `gutenberg-mobile` issues so could not test the final changes~~.  

### Prerequisites

1. Create and publish a long post (copy content from this [txt file](https://cloudup.com/iX9yC9byRyN)) in one of your test sites using Calypso.

2. Check out this branch.

3. Apply [this patch](https://cloudup.com/iA4hNRd2XBe) to add logs.

5. Install and run the build.

### To Test

**Case 1: Published Post + **Save on Exit** using **Update** button** 

1. Open the long post created above in the Gutenberg editor.
2. Change the title of the post by typing a few letters (make sure you remember the letters in the new title)
3. Quickly click the **Update** button on the editor.
4. Wait for the editor to exit and the post to be updated.
5. Notice in the logs that:
	a. [AutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L13-L17) exception (if gets logged) is a case of a false positive as we're preventing "auto-saves" after "user-initiated" save.
	b. [NoAutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L23-L26) exception (if gets logged) is a case of a false positive as the  “user-initiated” save actually had changes to be saved after a previous “auto-save” state. (See [previously ran test results](https://cloudup.com/iNf7Khi8bJm) with matched hash-code logs.)

Repeat same tests for the below cases: 

Case 2:  Published Post + **Save on Exit** using **Back** button 
Case 3:  Draft post + **Save on Exit** using **Publish** button 
Case 4:  Draft post + **Save on Exit** using **Back** button 

## Review/ Merge Instructions

Only 1 reviewer is sufficient.

1. ~~Make sure https://github.com/wordpress-mobile/WordPress-Android/pull/15508 is merged to develop.~~
2. Review this PR.
3. Merge  [issue/15509-remove-save-on-exit-exception](https://github.com/wordpress-mobile/WordPress-Android/tree/issue/15509-remove-save-on-exit-exception) and review changes.
4. Remove "Not Ready for Merge" label.
5. Merge the PR.

## Regression Notes
1. Potential unintended areas of impact
All flows which save a post on Editor exit.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested different cases manually (see above). 

3. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt [hunk 1]
```diff
@@ -57,6 +56,8 @@ class StorePostViewModel
     private val _onFinish = MutableLiveData<Event<ActivityFinishState>>()
     val onFinish: LiveData<Event<ActivityFinishState>> = _onFinish
```
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt:58]: I wonder if we should make this volatile.. wdyt?

```suggestion

    @Volatile
```

Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt:58]: Suggestion accepted 👍

**Comment by mkevins** [comment]
I've tested the cases described, and confirmed that this fixes cases 1 and 2: I reproduced the false positive crashes on `develop` and the same steps do not exhibit a crash with these changes (and the content is properly updated and saved).

For cases 3 and 4 still exhibited the crash, however:

<details>
<summary>Case 3</summary>

```
    --------- beginning of crash
2021-11-09 20:10:29.070 16857-16857/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
2021-11-09 20:10:29.070 16857-16857/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android.prealpha, PID: 16857
2021-11-09 20:10:29.070 16857-16857/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android.prealpha, PID: 16857
    Process: org.wordpress.android.prealpha, PID: 16857
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.os.Handler.handleCallback(Handler.java:938)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
    
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
    
    
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
    
    --------- beginning of system
    
    --------- beginning of system
```
</details>

<details>
<summary>Case 4</summary>

```
   --------- beginning of crash
    Process: org.wordpress.android.prealpha, PID: 17438
2021-11-09 20:14:57.998 17438-17438/org.wordpress.android.prealpha I/Sentry: Invoking inner uncaught exception handler.
2021-11-09 20:14:58.001 17438-17438/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
2021-11-09 20:14:57.998 17438-17554/org.wordpress.android.prealpha D/Sentry: Disk flush envelope fired
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
    Process: org.wordpress.android.prealpha, PID: 17438
    
    --------- beginning of crash
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
2021-11-09 20:14:58.001 17438-17438/org.wordpress.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
    Process: org.wordpress.android.prealpha, PID: 17438
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
    java.lang.RuntimeException: Debug build crash: Post had changes that needed to be saved when exiting the editor, and there was NOT an autosave pending. This means that the editor's autosave mechanism failed. This only crashes on debug builds.
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsyncOnEditorExit$20$EditPostActivity(EditPostActivity.java:1766)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$qBjDN7oKaBe_5-LzFaYOhW6ByM4.onPostUpdatedFromUI(Unknown Source:6)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at org.wordpress.android.ui.posts.EditPostActivity.lambda$updateAndSavePostAsync$19$EditPostActivity(EditPostActivity.java:1731)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at org.wordpress.android.ui.posts.-$$Lambda$EditPostActivity$dHwSEOESbDz01VTMYYNvc6SCVNQ.invoke(Unknown Source:8)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.Handler.handleCallback(Handler.java:938)
        at org.wordpress.android.ui.posts.EditPostRepository$updateAsync$1.invokeSuspend(EditPostRepository.kt:128)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at java.lang.reflect.Method.invoke(Native Method)
        at android.os.Looper.loop(Looper.java:223)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at android.os.Handler.handleCallback(Handler.java:938)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at android.os.Looper.loop(Looper.java:223)
```
</details>

Is this expected, or do we need to add the flag toggle in another code path for the draft posts? :thinking:

**Comment by ashiagr** [comment]
Hey @mkevins 👋 

Thank you so much for testing it 🙏 . In my testing, this was an expected outcome as the “user-initiated” save actually had changes to be saved after a previous “auto-save” state (note that as per our observations earlier `isAutosavePending` does not represent the true completion state for the concurrent auto-saving jobs). I also linked my [test results](https://cloudup.com/iNf7Khi8bJm) in the PR description. This crash can appear even for cases 1 and 2.

As long as we're sure that `title`/ `content` is serialized, and the current UI state (the source of truth) is getting saved, we do not need this exception logging. 

I can jump on a call to explain it further. Let me know. 🙂

**Review by mkevins — APPROVED** [review]
Thanks for clarifying Ashita :smile:

This, along with the changes in https://github.com/wordpress-mobile/WordPress-Android/commit/9ee6c4fe6ec7cef39844234834075590983e238c look good to merge together. I've confirmed that these are false positives, and we can remove these exceptions now. :+1:

**Pull request merged**